api/api.ts

```diff
@@ -1,8 +1,90 @@
-import axios from 'axios';
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import axios, { AxiosError, AxiosPromise, AxiosResponse } from 'axios';
+import { ApiPromise, ApiResponse, EmptyPayload, StatusMessageType } from '../types/apiResponse';
 
-const api = axios.create({
-  baseURL: '/api/',
-  timeout: 10 * 1000, // 10 seconds
-});
+class BaseApi {
+  client = axios.create({
+    baseURL: '/api/',
+    timeout: 10 * 1000, // 10 seconds
+  });
+
+  private getData<D>(url: string, params?: any): AxiosPromise<ApiResponse<D>> {
+    return this.client.get(url, params);
+  }
+
+  private postData<D>(url: string, data: any = {}): AxiosPromise<ApiResponse<D>> {
+    return this.client.post(url, data);
+  }
+
+  private patchData<D>(url: string, data: any = {}): AxiosPromise<ApiResponse<D>> {
+    return this.client.patch(url, data);
+  }
+
+  private deleteData<D>(url: string): AxiosPromise<ApiResponse<D>> {
+    return this.client.delete(url);
+  }
+
+  public get<D>(url: string, params?: any): ApiPromise<D> {
+    return processRequest(url, this.getData(url, params));
+  }
+
+  public post<D>(url: string, data: any = {}): ApiPromise<D> {
+    return processRequest(url, this.postData(url, data));
+  }
+
+  public patch<D>(url: string, data: any = {}): ApiPromise<D> {
+    return processRequest(url, this.patchData(url, data));
+  }
+
+  public delete<D>(url: string): ApiPromise<D> {
+    return processRequest(url, this.deleteData(url));
+  }
+}
+
+function processRequest<D>(endpoint: string, promise: AxiosPromise<ApiResponse<D>>): ApiPromise<D> {
+  return promise
+    .then((response: AxiosResponse) => {
+      const apiResponse = response.data;
+
+      if (process.env.NODE_ENV === 'development') {
+        console.info(`[API] ${response.status} ${endpoint} : ${getResponseMessages(apiResponse)}`);
+      }
+
+      return apiResponse;
+    })
+    .catch((error: AxiosError) => {
+      if (process.env.NODE_ENV === 'development') {
+        console.error(`[API] ${error.code} ${endpoint} : ${error.message}`);
+      }
+
+      throw makeApiErrorResponse(error);
+    });
+}
+
+function makeApiErrorResponse(error: AxiosError): ApiResponse<EmptyPayload> {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
```

Comment:
Can you briefly explain why we need to use `ts-ignore` for this and in line 80?

Reply:
Typescript doesn't know that our data has a messages field. Do you have any better solutions?

Reply:
See new thread

Reply:
Doesn't this work?: https://github.com/sivayogasubramanian/caza/pull/43#discussion_r970819855

Reply:
Doesn't this work: https://github.com/sivayogasubramanian/caza/pull/43#discussion_r970819855

Reply:
Sorry, I missed out on your comment earlier.

Yup this works, but by principle the argument is actually not of that type. The argument that we pass is actually of `AxiosError<ApiResponse<D>>`, but the type parameter was lost because it was written as `.catch((error: AxiosError)` on line 55.